hail/python/hailtop/aiotools/stream.py

```diff
@@ -203,3 +219,50 @@ def async_writable_blocking_readable_stream_pair() -> Tuple[AsyncQueueWritableSt
     blocking_readable = BlockingQueueReadableStream(q)
     async_writable = AsyncQueueWritableStream(q, blocking_readable)
     return async_writable, blocking_readable
+
+
+class BlockingCollect(_Closable):
```

Comment:
Why is this called `Blocking Collect`? It seems like the difference between this and BlockingQueueReadableStream is that one reads bytes into a byte array and this new class returns bytes in a new byte array.

Reply:
Hmm. Good question. I think you've identified the essential difference. BlockingQueueReadableStream is a ReadableStream (built on a BlockingQueue). BlockingCollect does not implement the ReadableStream API. Indeed, it seems to be a "sink" (or, I guess, "collector") which simply reads the entire stream into a byte array.

IIUC, this is only used in the S3 API for piecewise ("CreatePart") object upload. In that setting, we break a file into many pieces and upload each piece to S3 in a separate request. The S3 API requires a byte or byte array, not a stream. Therefore, we "collect" the entire file-piece into one bytearray which is retrieved by S3CreatePartMAnager with `.get()`.